### PR TITLE
Convert unison to use optparse-applicative

### PR DESCRIPTION
**Pull request by shmish111**
## Overview

This has a number of benefits that come at the expence of some flexibility in how the help messages are displayed and the structure of the commands.

* The code in `Main.main` is more well typed with a lot less strings
* More logic has been moved into the options parsing such as the number of files provided to unison transcript making `runTranscripts` simpler and more focused
* The `init` command was missing from the help, this type of bug is less likely to happen as the data type and structure are more visible in Options.hs
* The limitations on how the CLI is structured lead to a more standard way of doing things
* Free shell completion scripts with https://github.com/pcapriotti/optparse-applicative#bash-zsh and-fish-completions

Note that optparse-applicative provides plenty of options for formatting including native use of http://hackage.haskell.org/package/ansi-wl-pprint so it should be possible to do _almost_ anything that's already done.

I expect some changes would need to be made to the help text on this PR.

## Implementation notes

* Move CLI parsing logic into its own module
* Treat the CLI input as a type rather than a list of strings
* I made a few refactorings
* I formatted `Main.hs` and `Options.hs` using https://hackage.haskell.org/package/ormolu

## Interesting/controversial decisions

This is possibly a controversial PR in general since it sacrifices flexibility in the help output and trades off flexibility in the CLI structure

## Test coverage

No extra tests added, debatable as to whether tests would be a good idea or not for this type of thing.

## Loose ends

None that I can think of

### CODE REVIEW

**Comment by pchiusano** [comment]
@shmish111 cool! could you include some screenshots of the new help output? And maybe some screen shots of a couple error cases? (Very helpful when reviewing these kinds of user-facing PRs)

**Comment by shmish111** [comment]
The base command remains unchanged:
```
➜  unison git:(optparse) unison

   _____     _
  |  |  |___|_|___ ___ ___
  |  |  |   | |_ -| . |   |
  |_____|_|_|_|___|___|_|_|

  Welcome to Unison!

  I'm currently watching for changes to .u files under ~/personal/unison

  Type help to get help. 😎

.> quit
```
Basic help is a lot smaller but more structured. You could add a nice big description if required
```
➜  unison git:(optparse) unison --help
Usage: unison [--codebase ARG] [COMMAND]

Available options:
  --codebase ARG           The path to the codebase, defaults to the home
                           directory
  -h,--help                Show this help text

Available commands:
  version                  print the version of unison
  init                     Initialise a unison codebase
  run                      Execute a definition
  transcript               Execute transcript markdown files
```
A simple command
```
➜  unison git:(optparse) unison version
ucm version: release/M2-base-1048-gb876c9e7'
```
Help for a sub-command
```
➜  unison git:(optparse) unison init --help
Usage: unison init
  Initialise a unison codebase

Available options:
  -h,--help                Show this help text
```
A more complicated example
```
➜  unison git:(optparse) unison transcript --help
Usage: unison transcript [--fork] [--save-codebase] transcriptfiles...
  Execute transcript markdown files

Available options:
  --fork                   if set the transcript is executed in a copy of the
                           current codebase
  --save-codebase          if set the resulting codebase will be saved to a new
                           directory, otherwise it will be deleted
  -h,--help                Show this help text

For each transcript file provided this executes the transcript and creates `mytranscript.output.md` if successful.
Exits after completion, and deletes the temporary directory created.
Multiple transcript files may be provided; they are processed in sequence starting from the same codebase.
```
Lets say we forget to supply input files for `unison transcript`
```
➜  unison git:(optparse) unison transcript
Missing: transcriptfiles...

Usage: unison transcript [--fork] [--save-codebase] transcriptfiles...
  Execute transcript markdown files

Available options:
  --fork                   if set the transcript is executed in a copy of the
                           current codebase
  --save-codebase          if set the resulting codebase will be saved to a new
                           directory, otherwise it will be deleted
  -h,--help                Show this help text

For each transcript file provided this executes the transcript and creates `mytranscript.output.md` if successful.
Exits after completion, and deletes the temporary directory created.
Multiple transcript files may be provided; they are processed in sequence starting from the same codebase.
```
What about errors that are not in the options parsing logic such as a missing file
```
➜  unison git:(optparse) unison --codebase ../unison-playground
No codebase exists in /Users/davidsmith/personal/unison-playground
Run `ucm -codebase /Users/davidsmith/personal/unison-playground init` to create one, then try again!
```
Lets do what it says!
```
➜  unison git:(optparse) unison --codebase ../unison-playground init
Initializing a new codebase in: /Users/davidsmith/personal/unison-playground
```
And now we can use a different codebase location
```
➜  unison git:(optparse) unison --codebase ../unison-playground

   _____     _
  |  |  |___|_|___ ___ ___
  |  |  |   | |_ -| . |   |
  |_____|_|_|_|___|___|_|_|

  Welcome to Unison!

  I'm currently watching for changes to .u files under ~/personal/unison

  Type help to get help. 😎

  Enter pull https://github.com/unisonweb/base .base to set up the default base library. 🏗

.>
```

**Comment by pchiusano** [comment]
I think I'm sold! Going to sleep on it and will give more of a review later. @runarorama or @aryairani lmk if you have any strong feelings on this one.

**Comment by pchiusano** [comment]
Also @shmish111 in the PR can you add self to CONTRIBUTORS file if you haven't already?

File: parser-typechecker/unison/Options.hs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        cmd :: Command
+      }
+  deriving (Show)
+
+data Command
+  = Launch
+  | Version
+  | Init
+  | Run (Maybe FilePath) Bool String
+  | Transcript Bool Bool [FilePath]
```
Comment (aryairani) [inline parser-typechecker/unison/Options.hs:38]: Could use a comment to understand these args inline

**Comment by shmish111** [comment]
thanks @aryairani 

I can fix everything in your list _except_ 

> I was a little surprised that unison --help transcript gave the same output as unison --help — I needed to have put them in the other order: unison transcript --help.

That's a bit of an edge case because `--help` is command specific, so it's _slightly_ unclear what that should do. Strictly speaking I think you could do it but you would have to re-write all of the help helpers and the help generation yourself, which kind of destroys the point of it. I suggest that if I fix everything else then this is a small enough problem to ignore? I also raised https://github.com/pcapriotti/optparse-applicative/issues/379 for this issue.

File: parser-typechecker/unison/Main.hs [hunk 2]
```diff
@@ -1,225 +1,159 @@
...
       launch currentDir configFilePath theCodebase []
-    [version] | isFlag "version" version ->
-      putStrLn $ "ucm version: " ++ Version.gitDescribe
-    [help] | isFlag "help" help -> PT.putPrettyLn usage
-    ["init"] -> FileCodebase.initCodebaseAndExit mcodepath
-    "run" : [mainName] -> do
-      theCodebase <- FileCodebase.getCodebaseOrExit mcodepath
+    Version -> putStrLn $ "ucm version: " <> Version.gitDescribe
+    Init -> FileCodebase.initCodebaseAndExit codepath
+    Run _ True mainName -> do
```
[the remark this answers is not in the document]
Comment (shmish111) [reply parser-typechecker/unison/Main.hs]: Added newtypes to make things clearer

File: parser-typechecker/unison/Main.hs [hunk 3]
```diff
@@ -1,225 +1,159 @@
...
-    _ -> do
-      PT.putPrettyLn usage
-      Exit.exitWith (Exit.ExitFailure 1)
+    Run _ _ _ -> Exit.die "Expected a unison file with extension .u"
+    Transcript fork save transcripts -> runTranscripts fork save currentDir codepath transcripts
 
-prepareTranscriptDir :: Bool -> Maybe FilePath -> IO FilePath
-prepareTranscriptDir inFork mcodepath = do
-  currentDir <- getCurrentDirectory
+prepareTranscriptDir :: Bool -> FilePath -> Maybe FilePath -> IO FilePath
```
[the remark this answers is not in the document]
Comment (shmish111) [reply parser-typechecker/unison/Main.hs]: newtypes again. indeed `False` + `Some path` just results in the path being ignored

File: parser-typechecker/unison/Main.hs [hunk 4]
```diff
@@ -1,225 +1,159 @@
...
+              ]
+          )
+
+runTranscripts :: Bool -> Bool -> FilePath -> Maybe FilePath -> [String] -> IO ()
+runTranscripts inFork keepTemp currentDir mcodepath args = do
+  transcriptDir <- prepareTranscriptDir inFork currentDir mcodepath
+  runTranscripts' currentDir mcodepath transcriptDir args
+  case keepTemp of
+    True -> removeDirectoryRecursive transcriptDir
+    False ->
```
[the remark this answers is not in the document]
Comment (shmish111) [reply parser-typechecker/unison/Main.hs]: oh dear, well spotted

File: parser-typechecker/unison/Options.hs (hunk 1 above)
Comment (shmish111) [reply parser-typechecker/unison/Options.hs]: newtypes!

**Comment by shmish111** [comment]
Ok I've pushed a few more commits that I think should fix everything mentioned by @aryairani except the `--help` order issue I mentioned above.

> It looks like Text.PrettyPrint.ANSI.Leijen locks the wrapping width at 80 columns

It turns out its kind of the opposite of that. If you use the `xxxDoc` family of functions this means you can provide a Leijin `Doc` rather than a `String` and when you do this it uses the full width (or you can set the width explicitly).

Overall the help text can be fiddled about with ad-infinitum and it does currently feel a little light however at this point I don't think I have enough knowledge about the product to do much more than I have.

**Comment by aryairani** [comment]
Sorry for the delay, just getting back to reviewing PRs after getting out from under mine.

**Comment by aryairani** [comment]
I definitely like the better organization, definitely like the new Options data type; newtypes, and like having the parsing tidily out of the way, etc.

I also would like to merge it before it gets too out of date, but I'm stuck in a tough spot between "unconvinced re. the help generation", and "don't have time/priority right now to learn on my own about the different customization possibilities and take over maintenance". If trying to customize the help generation, destroys the point of it—I wasn't sure about this, wanted to get your opinion again—then I'm not sure how to proceed.  WDYT?

**Comment by shmish111** [comment]
Personally I think that the help generation isn't better or worse than it was before, just different. Given the benefits of having something structured and standard I feel it makes sense to use it until the time the help generation isn't good enough. At that time it would then be worth making something custom to do the same job. 

I guess the problem is that CLI UX can easily become a bit of a rabbit hole and I would argue that unless you have some very specific UX in mind then going with the standard library and UX (which in the case of Haskell is optparse and Leijen) is the way to go. Especially if your concern is that you don't have resources to spend on it.

But... I opened this PR so I'm maybe not looking at this 100% objectively.

**Pull request closed without merging**